### PR TITLE
8/decouple

### PR DESCRIPTION
**Pull request by joel99**
String hacking. Improvement would be for `PluginMaster` and `PluginReturn` entries be inserted in place but this currently works. 
Reasoning: Alternatives: `merge-graphql-schemas` throws errors on merging inputs, build in config doesn't resolve top-level conflicts (i.e. PluginMaster). Our plugin system is guaranteed to not conflict (directory can only have one file of the same name).

### CODE REVIEW

File: server/app.ts [hunk 1]
```diff
@@ -40,14 +38,20 @@ const resolvers = {
 
 					return async () => { // Sending function
 						try {
-							statusRet[name] = await plugin.sendMessage(message, verifiedConfig);
+							statusRet.push({
```
Comment (illegalprime) [inline server/app.ts]: Instead of pushing you can just return the object, and then when you call `Promise.all` it'll give you an array.

File: server/typeDefs.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+const mainTypeDefs = fs.readFileSync(path.resolve(__dirname, "../api.graphql"), "utf8");
+const typeDefFileArr = fs.readdirSync(path.resolve(__dirname, "../server/plugins/graphql"), "utf8");
+const rawPluginTypeDefs = typeDefFileArr.map(fn => fs.readFileSync(path.resolve(__dirname, "../server/plugins/graphql", fn), "utf8"));
+
+// E.g. Slack
+const baseNames = typeDefFileArr.map(fn => fn.substring(0, fn.indexOf("Plugin.graphql")));
+
+// Generate PluginMaster body (See base api.graphql)
+let pluginMasterStr = "\ninput PluginMaster {";
```
Comment (illegalprime) [inline server/typeDefs.ts]: Maybe you can use templating here (if you think that's more readable):
```typescript
const pluginMasterStr = `
input PluginMaster {
  ${stuff}
};
`
```

File: server/typeDefs.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+const configStrArr = baseNames.map(name => `${name.toLowerCase()}: ${name}Config`);
+configStrArr.forEach(s => {
```
Comment (illegalprime) [inline server/typeDefs.ts]: you should just `configStrArr.join("\n\t")`

File: server/typeDefs.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	pluginMasterStr += "\n\t" + s;
+});
+pluginMasterStr = pluginMasterStr + "\n}\n\n"; // Formatting
+
+// Append pluginTypeDefs (rename each Config)
+// IE input Config => input SlackConfig
+const processedPluginTypeDefs = rawPluginTypeDefs.map((s, i) => s.slice(0, 6) + baseNames[i] + s.slice(6));
```
Comment (illegalprime) [inline server/typeDefs.ts]: Again, super brittle code here, maybe just have each plugin export:
```typescript
interface Notifier<T> {
  schema(): string {
    return `{
      _: Boolean
    }`;
  }
}
```
As in have the plugin not even name its own schema, so no `input SlackConfig` here.

File: server/typeDefs.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+let processedStr = "";
+processedPluginTypeDefs.forEach(s => {
```
Comment (illegalprime) [inline server/typeDefs.ts]: `.join("\n")` here too. (I don't know about javascript, but in other languages this is often more efficient too!)

File: server/typeDefs.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+	processedStr += s + "\n";
+});
+processedStr = processedStr.substring(0, processedStr.length - 1); // Formatting
+
+const mergedTypeDefs = mainTypeDefs + pluginMasterStr + processedStr;
```
Comment (illegalprime) [inline server/typeDefs.ts]: Templating here, e.g.
```typescript
const merged = `
${mainTypeDefs}
${pluginMasterStr}
${processedStr}
`;
```
or `.join("\n")`, but just using `+` might mean you forget to include a space to separate the three strings and then the thing breaks.

File: server/plugins/Twitter.ts [hunk 7]
```diff
@@ -57,28 +61,28 @@ export default class Twitter implements GenericNotifier<{}> {
 		const params = { status: message };
 		const res: Promise<APIReturn> = new Promise((resolve, reject) => {
 			client.post('statuses/update', params, (error, data: { user?: { screen_name?: string }; id_str?: string }, response) => {
-				if (!error) {
```
Comment (illegalprime) [inline server/plugins/Twitter.ts]: What's with the indentation changes?

File: package.json [hunk 8]
```diff
@@ -26,9 +26,10 @@
     "compression": "^1.7.1",
     "dotenv": "^5.0.1",
     "express": "^4.16.2",
-    "graphql": "^0.12.3",
+    "graphql": "^0.13.0",
     "graphql-server-express": "^1.3.2",
     "graphql-tools": "^2.21.0",
+    "merge-graphql-schemas": "^1.5.0",
```
Comment (illegalprime) [inline package.json]: remove this if your not using it

File: server/plugins/Twitter.ts [hunk 9]
```diff
@@ -5,6 +5,10 @@ import * as TwitterAPI from "twit";
 
 export default class Twitter implements GenericNotifier<{}> {
 
+	public schema: string = `{
+\t_: Boolean
```
Comment (illegalprime) [inline server/plugins/Twitter.ts]: Don't worry about putting `\t` here or making the indent perfect. I would remove `\t` and just indent everything so it looks nice, GraphQL won't care.

File: server/typeDefs.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+});
+const pluginMasterBody = configStrArr.join("\n\t");
+const pluginMasterStr = `input PluginMaster {\n\t${pluginMasterBody}\n}`;
+
+const processedPluginTypeDefs = Object.keys(plugins.mediaAPI).map(plugin => `input ${plugin}Config ${plugins.mediaAPI[plugin].schema}`);
+const processedStr = processedPluginTypeDefs.join("\n\n");
+
+const mergedTypeDefs = `
+${mainTypeDefs}
+${pluginMasterStr}\n
```
Comment (illegalprime) [inline server/typeDefs.ts]: Don't think you need the `\n` here.

File: server/app.ts [hunk 11]
```diff
@@ -21,47 +18,52 @@ process.on("unhandledRejection", err => {
...
 
-			const checkQueue = Object.keys(args.plugins).map( name => {
+			const checkQueue = Object.keys(args.plugins).map( rawName => {
 
 				return (async () => { // Loading checkQueue IIFE
-
-					const plugin: GenericNotifier<any> = plugins.mediaAPI[name.toUpperCase()];
-					const verifiedConfig = await plugin.check(args.plugins[name]); // Verify
+					// Upper Cameled
+					const name = rawName.split('_').map(w => w[0].toUpperCase() + w.slice(1)).join('');
```
Comment (illegalprime) [inline server/app.ts]: This makes me nervous about keeping this implementation in sync with the generation code, and making sure one is the reverse of the other. A few options here:
1. Write some tests to make sure that these conversions work (so they don't become out of sync in the future)
2. Don't make this conversion, for a plugin like `LiveSite` just have the type named `LiveSiteConfig` and the key for it be named `livesite` so you just have to do a `.toLowerCase()` and search through the map (or maybe make another map with the lowercased values)

File: server/typeDefs.ts (hunk 10 above)
Comment (joel99) [reply server/typeDefs.ts]: It's for an extra line of space, I checked console.log

Comment (illegalprime) [reply server/typeDefs.ts]: :+1:

**Comment by illegalprime** [comment]
Everything's good! Just the GraphQL merging test and adding some test runner (`mocha`) is needed.

File: server/test/stringTests.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+	if (actual !== after) console.log(`ERROR: expected "${after}".`);
+	return actual === after;
+};
+
+function main() {
+	console.log("Running tests...");
+	// Upper camel has less worry about edge cases since graphql is machine generated
+	testUpperCamel("", "");
+	testUpperCamel("live", "Live");
+	testUpperCamel("LiveSite", "LiveSite");
```
Comment (illegalprime) [inline server/test/stringTests.ts]: When you use expect and other fancy tools this should just be like:
```typescript
upperCamel("LiveSite").equals("LiveSite");
```
or something similar (I don't remember the syntax).
